src/uct/cuda/cuda_ipc/cuda_ipc_cache.c

```diff
@@ -183,230 +138,172 @@ uct_cuda_ipc_get_remote_cache(pid_t pid, uct_cuda_ipc_cache_t **cache)
 {
-    ucs_status_t status = UCS_OK;
-    uct_cuda_ipc_cache_t *cache;
-    ucs_pgt_region_t *pgt_region;
-    uct_cuda_ipc_cache_region_t *region;
-
-    status = uct_cuda_ipc_get_remote_cache(pid, &cache);
-    if (status != UCS_OK) {
-        return status;
-    }
-
-    /* use write lock because cache maybe modified */
-    pthread_rwlock_wrlock(&cache->lock);
-    pgt_region = UCS_PROFILE_CALL(ucs_pgtable_lookup, &cache->pgtable, d_bptr);
-    ucs_assert(pgt_region != NULL);
-    region = ucs_derived_of(pgt_region, uct_cuda_ipc_cache_region_t);
+    cuIpcCloseMemHandle((CUdeviceptr)mapped_addr);
+}
 
-    ucs_assert(region->refcount >= 1);
-    region->refcount--;
+ucs_status_t
+uct_cuda_ipc_unmap_memhandle(uct_cuda_ipc_md_t *md,
+                             const uct_cuda_ipc_key_t *key,
+                             void *mapped_addr,
+                             uct_cuda_ipc_rcache_region_t *cuda_ipc_region)
+{
+    ucs_status_t status;
+    ucs_rcache_t *cache;
 
-    /*
-     * check refcount to see if an in-flight transfer is using the same mapping
-     */
-    if (!region->refcount && !cache_enabled) {
-        status = ucs_pgtable_remove(&cache->pgtable, &region->super);
+    if (md->rcache_enable != UCS_NO) {
+        status = uct_cuda_ipc_get_remote_cache(md, key->pid, &cache);
         if (status != UCS_OK) {
-            ucs_error("failed to remove address:%p from cache (%s)",
-                      (void *)region->key.d_bptr, ucs_status_string(status));
+            return status;
         }
-        ucs_assert(region->mapped_addr == mapped_addr);
-        status = UCT_CUDADRV_FUNC_LOG_ERR(
-                cuIpcCloseMemHandle((CUdeviceptr)region->mapped_addr));
-        ucs_free(region);
+
+        ucs_rcache_region_put(cache, &cuda_ipc_region->super);
+    } else {
+        uct_cuda_ipc_close_memhandle(mapped_addr, NULL);
     }
 
-    pthread_rwlock_unlock(&cache->lock);
-    return status;
+    return UCS_OK;
 }
 
-UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_ipc_map_memhandle, (key, mapped_addr),
-                 const uct_cuda_ipc_key_t *key, void **mapped_addr)
+ucs_status_t uct_cuda_ipc_open_memhandle(void **mapped_addr,
+                                         const uct_cuda_ipc_key_t *key,
+                                         uct_cuda_ipc_rcache_region_t *cuda_ipc_region,
+                                         ucs_rcache_t *rcache)
 {
-    uct_cuda_ipc_cache_t *cache;
-    ucs_status_t status;
-    ucs_pgt_region_t *pgt_region;
-    uct_cuda_ipc_cache_region_t *region;
-    int ret;
+    CUresult cuerr;
+    const char *cu_err_str;
 
-    status = uct_cuda_ipc_get_remote_cache(key->pid, &cache);
-    if (status != UCS_OK) {
-        return status;
+    cuerr = cuIpcOpenMemHandle((CUdeviceptr *)mapped_addr,
+                               key->ph, CU_IPC_MEM_LAZY_ENABLE_PEER_ACCESS);
+    if ((cuerr == CUDA_SUCCESS) || (cuerr == CUDA_ERROR_ALREADY_MAPPED)) {
+        if (cuda_ipc_region != NULL) {
```

Comment:
I would expect we'd always have to return the mapped handle from this function

Reply:
This function has to always return the mapped address (`mapped_addr`) but doesn't have to always populate cuda_ipc_region. We don't need to when rcache is disabled.

Reply:
But the caller already knows key->ph. So it could be set outside this function.